### PR TITLE
automatically activate .condaenv environments; fixes for reticulate binding to Anaconda

### PR DESCRIPTION
**Pull request by kevinushey**
### Intent

This PR contains two main changes:

1. If the project contains a Conda environment within the `.condaenv` directory, RStudio will now automatically activate that environment on startup.

2. Add some extra glue to ensure that `reticulate` can bind to Anaconda installations on macOS. See https://github.com/ContinuumIO/anaconda-issues/issues/12478 for more details.

### Approach

The Anaconda glue is the more interesting part of the PR. Currently, Anaconda installations rely on the process rpath when discovering libraries. This unfortunately doesn't work for us, since we attempt to load Python as a library (and so don't have the rpath entries set that Anaconda is expecting).

The workaround is to set `DYLD_FALLBACK_LIBRARY_PATH` to include the Anaconda library path, but this is challenging for a couple reasons:

1. The `DYLD_FALLBACK_LIBRARY_PATH` cannot be changed after a process has already been launched. (You can change the value of the environment variable, but it will not affect future library load attempts in the current process.)
2. When RStudio launches the `rsession` process, we have no idea if the user plans to use Python, or even Anaconda Python. We also don't know where Anaconda might be installed.

The solution is to instead have RStudio place a special fallback library path in the `DYLD_FALLBACK_LIBRARY_PATH`, and have R initialize the library path when necessary. See https://github.com/rstudio/reticulate/commit/08f145b0d2f13c8f8c242ea788359406d139467d for an example.

### Automated Tests

N/A

### QA Notes

Verify, on macOS, that the following works:

1. Install Anaconda: https://www.anaconda.com/products/individual#macos
2. Install the development version of `reticulate`: `remotes::install_github("rstudio/reticulate")`
3. Try to load that environment: `reticulate::use_python("~/opt/anaconda3/bin/python", required = TRUE); reticulate::py_config()`

If all is well, you should see `reticulate::py_config()` print without issue.

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

File: src/cpp/session/SessionMain.cpp [hunk 1]
```diff
@@ -239,6 +239,20 @@ using namespace rsession::client_events;
 namespace rstudio {
 namespace session {
 
+namespace {
+
+std::string s_fallbackLibraryPath;
+
+void onExit()
```
Comment (jjallaire) [inline src/cpp/session/SessionMain.cpp:246]: Can the value we set in our executable actually persist after we exit? (i.e. do we actually need this code)

File: src/cpp/session/modules/SessionPythonEnvironments.R [hunk 2]
```diff
@@ -43,11 +43,92 @@
...
+   # if this project has a local interpreter, use it
+   if (!is.null(projectDir))
+   {
+      projectPython <- .rs.python.projectInterpreterPath(projectDir)
+      if (file.exists(projectPython))
+         return(projectPython)
+   }
+   
+   # check version of Python configured by user
+   prefsPython <- .rs.readUiPref("python_path")
```
Comment (jjallaire) [inline src/cpp/session/modules/SessionPythonEnvironments.R:100]: Quick reminder (if it isn't already resolved) to make sure that we persist python_path that is a subdirectory of the project as a relative path.

**Comment by jjallaire** [comment]
Does the IDE prompt for reticulate installation when the user clicks the "Select Intepreter" button?

We should keep a note or file a separate issue that RStudio should require the CRAN version of reticulate that has the requisite changes.

**Comment by jjallaire** [comment]
This probably needs some updating to reflect this subsequent PR: https://github.com/rstudio/rstudio/pull/9510

I noticed that VS code will actually run the conda activation script in the terminal when configured to bind to a project level conda environment. How hard would it be for us to do this? (and/or is it even necessary)

**Comment by kevinushey** [comment]
> I noticed that VS code will actually run the conda activation script in the terminal when configured to bind to a project level conda environment. How hard would it be for us to do this? (and/or is it even necessary)

We can't really do this in the main session (or, doing it is quite challenging) since this implies running a command that modifies the process's current environment. That can only work if conda injects its own shims into the shell, so that running `conda activate` really runs a `__conda_activate` shell function that modifies the current environment. (It also implies that `conda activate` will fail if the user hasn't already called `conda init` or something similar.)

We might be able to do this in our Terminal instances, but we again run into the potential `conda init` fiasco.

Fortunately, we can basically do the same by manually setting the PATH and setting some environment variables (e.g. `CONDA_PREFIX`) to tell Conda to target a particular environment.

File: src/cpp/session/modules/SessionPythonEnvironments.R (hunk 2 above)
Comment (kevinushey) [reply src/cpp/session/modules/SessionPythonEnvironments.R:100]: Should be fixed up now -- thanks for spotting that.

**Pull request merged**